lib.sh

```diff
@@ -91,16 +93,63 @@ function get_package_name_ver {
 # Arguments:
 #   The comma and/or space delimited list of packages.
 # Returns:
-#   Sorted list of space delimited packages.
+#   Sorted list of space delimited package name=version pairs.
 ###############################################################################
 function get_normalized_package_list {
-  # Remove commas, and block scalar folded backslashes.
-  local stripped=$(echo "${1}" | sed 's/[,\]/ /g')
-  # Remove extraneous spaces at the middle, beginning, and end.
-  local trimmed="$(\
-    echo "${stripped}" \
-    | sed 's/\s\+/ /g; s/^\s\+//g; s/\s\+$//g')"
-  echo ${trimmed} | tr ' ' '\n' | sort | tr '\n' ' '
+  # Remove commas, and block scalar folded backslashes,
+  # extraneous spaces at the middle, beginning and end
+  # then sort.
+  packages=$(echo "${1}" | \
+    sed 's/[,\]/ /g; s/\s\+/ /g; s/^\s\+//g; s/\s\+$//g' | \
+    sort -t' ')
+
+  # Validate package names and get versions.
+  log_err "resolving package versions..."
+  data=$(apt-cache --quiet=0 --no-all-versions show ${packages} 2>&1 | \
+    grep -E '^(Package|Version|N):')
+  log_err "resolved"
+
+  local ORIG_IFS="${IFS}"
+  IFS=$'\n'
+  declare -A missing
+  local package_versions=''
+  local pkg='' sep=''
+  for v in ${data}; do
```

Comment:
Can we use a more descriptive name? Also, the iterative variable can be something like pair or key_value?